improver/calibration/reliability_calibration.py

```diff
@@ -585,6 +585,86 @@ def _ensure_monotonicity(self, cube):
             warnings.warn(msg)
             cube.data = np.sort(cube.data, axis=threshold_dim)
 
+    @staticmethod
+    def _sum_pairs(array, lower, upper):
+        """
+        Replace a pair of values in an array with their sum.
+
+        Args:
+            array (numpy.ndarray):
+                Array to be modified.
+            lower (int):
+                Lower index of pair.
+            upper (int):
+                Upper index of pair.
+
+        Returns:
+            numpy.ndarray:
+                Array where a pair of values has been replaced by their sum.
+        """
+        array[lower] = np.sum(array[lower:upper + 1])
+        return np.delete(array, upper)
+
+    def _combine_bin_pair_for_observation_frequency_monotonicity(
+        self, observation_count, forecast_probability_sum, forecast_count
+    ):
+        """
+        Combine a pair of bins when non-monotonicity of the observation
+        frequency is detected. Iterate top-down from the highest forecast
+        probability bin to the lowest probability bin when combining the bins.
+        Only allow a single pair of bins to be combined.
+
+        Args:
+            observation_count (numpy.ndarray):
+                Observation count extracted from reliability table.
+            forecast_probability_sum (numpy.ndarray):
+                Forecast probability sum extracted from reliability table.
+            forecast_count (numpy.ndarray):
+                Forecast count extracted from reliability table.
+
+        Returns:
+            Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]:
+                Tuple containing the updated observation count,
+                forecast probability sum and forecast count.
+
+        """
+        observation_frequency = np.array(observation_count / forecast_count)
+        for upper in np.arange(len(observation_frequency)-1, 0, -1):
+            lower = upper - 1
+            (diff,) = np.diff(
+                [observation_frequency[lower], observation_frequency[upper]]
+            )
+            if diff < 0:
+                forecast_count = self._sum_pairs(forecast_count, lower, upper)
+                observation_count = self._sum_pairs(observation_count, lower, upper)
+                forecast_probability_sum = self._sum_pairs(
+                    forecast_probability_sum, lower, upper
+                )
+                break
+        return observation_count, forecast_probability_sum, forecast_count
+
+    def _assume_constant_observation_frequency(self, observation_frequency):
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6001543)